pkg/services/ngalert/migration/service.go

```diff
@@ -64,31 +69,41 @@ func (ms *migrationService) Run(ctx context.Context) error {
 			if err != nil {
 				return fmt.Errorf("getting migration status: %w", err)
 			}
-			if migrated == ms.cfg.UnifiedAlerting.IsEnabled() {
-				// Nothing to do.
-				ms.log.Info("No migrations to run")
-				return nil
-			}
-
-			if migrated {
-				// If legacy alerting is also disabled, there is nothing to do
-				if setting.AlertingEnabled != nil && !*setting.AlertingEnabled {
+			if !ms.cfg.UnifiedAlerting.IsEnabled() {
+				//nolint:staticcheck
+				if migrated && ms.cfg.ForceMigration { // TODO: Remove in v11.
+					ms.log.Info("ForceMigration enabled, deleting unified alerting data")
+					err := ms.migrationStore.RevertAllOrgs(ctx)
+					if err != nil {
+						return fmt.Errorf("force_migration revert: %w", err)
+					}
+					ms.log.Info("Unified alerting data deleted")
 					return nil
 				}
 
-				// Safeguard to prevent data loss when reverting from UA to LA.
-				if !ms.cfg.ForceMigration {
-					return ForceMigrationError
+				// Set status to false so that next time UA is enabled, we run the migration again. That is when
+				// CleanUpgrade will be checked to determine if revert should happen.
+				err = ms.migrationStore.SetMigrated(ctx, anyOrg, false)
+				if err != nil {
+					return fmt.Errorf("setting migration status: %w", err)
 				}
+				return nil
+			}
+
+			if migrated {
+				ms.log.Info("Migration already run")
+				return nil
+			}
 
+			// Safeguard to prevent data loss.
+			if ms.cfg.UnifiedAlerting.Upgrade.CleanUpgrade {
+				ms.log.Info("CleanUpgrade enabled, deleting unified alerting data")
```

Comment:
I think this flag does not work now because there is no dry-run migration yet, and therefore it's not possible to achieve `migrated=false` and org migration is "true".

For uninformed readers, these lines could be confusing. Can you leave a message that `migrate` here is an indicator of a future "committed" migration, which can be false, and at the same time, there is migration info (produced by the migration helper app) that needs to be cleaned?

Reply:
Yes, I forgot about rollback!